File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(subRoot == nullptr)
            return true;
        if(root == nullptr)
            return false;
        return isSameTree(root, subRoot) ||
        isSubtree(root->left, subRoot) ||
        isSubtree(root->right, subRoot);
    }

    bool isSameTree(TreeNode* rootLeft, TreeNode* rootRight) {
        int a = (rootLeft == nullptr) ? -1 : rootLeft->val;
        int b = (rootRight == nullptr) ? -1 : rootRight->val;
        // cout << a << ' ' << b << ' ';
        if(rootLeft == rootRight) {
            // cout << "true\n";
            return true;
        } else {
            if (rootLeft == nullptr || rootRight == nullptr) {
                // cout << "false\n";
                return false;
            } else {
                bool res = (rootLeft->val == rootRight->val) && 
                isSameTree(rootLeft->left, rootRight->left) && 
                isSameTree(rootLeft->right, rootRight->right);
                // cout << ((res == true) ? "true" : "false") << '\n';
                return res;
            }
        }
    }
};

File: Data Structures & Algorithms/depth-of-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # utilizando BFS em vez de DFS
        level = 0
        node_queue = deque()
        if root:
            node_queue.append(root)
        print(node_queue)
        while node_queue:
            level += 1
            
            for i in range(len(node_queue)):
                root = node_queue.popleft()
                if root and root.left:
                    node_queue.append(root.left)
                if root and root.right:
                    node_queue.append(root.right)
            

        return level